data/blog/2024/02/generic-fuzzy-search-algorithm.md

```diff
@@ -0,0 +1,293 @@
+	return FuzzySearchString(characters: searchableString.normalise())
+}
+```
+This function returns the normalised version of the searchableString that is provided by the object that it is called from.
+```swift
+func fuzzyMatch(query: String) -> FuzzySearchMatchResult {
+	let characters = normaliseString()
+	
+	return fuzzyMatch(query: query, characters: characters)
+}
+```
+Subsequently, we simply transform the characters into their normalised counterparts and invoke the fuzzyMatch function. This, in turn, provides us with the `FuzzySearchMatchResult`.
+
+### Fuzzy Search Extension
+Now, let's make arrays fuzzy-searchable by writing an extension for them. This way, they can easily be searched using our fuzzy match functionality.
+```swift
+extension Collection where Iterator.Element: FuzzySearchable {
+	func fuzzySearch(query: String) -> [(result: FuzzySearchmatchResult), item: Iterator.Element)] {
+		return map {
+			(result: $0.fuzzyMatch(query: query), item: $0)
+		}.filter {
+			$0.result.weight > 0
+		}.sorted {
+			$0.result.weight > $1.result.weight
+		}
+	}
+}
+```
+To enable fuzzy searching, we extend Swift's `Collection` with a handy `fuzzySearch` method. However, to ensure compatibility, we limit this extension to collections whose elements conform to the `FuzzySearchable` protocol we defined earlier.
+When using this method, you simply provide a search `query`. The function then iterates over the array, invoking the `fuzzyMatch` method on each element. The `fuzzyMatch` method, as you recall, provides a weight for every element. We leverage these weights to filter out elements with a score less than or equal to 0. Finally, we sort the array, placing the element with the highest score at the top.
+*Additionally, for a slight performance boost, consider utilising a `concurrentMap` from the `CollectionConcurrencyKit` package.*
+
+### Usage
+In CodeEdit we for one use the algorithm to provide a quick open overlay that lets user find files quickly. To apply fuzzy search to an array of URLs, you can easily extend the data type `URL` to conform to the `FuzzySearchable` protocol.
+```swift
+extension URL: FuzzySearchable {
+    var searchableString: String {
+        return self.lastPathComponent
+    }
+}
+```
+
+If you've created your custom data type, making it compatible with fuzzy search is a breeze—simply ensure it conforms to the `FuzzySearchable` protocol, as shown here:
+```swift
+struct Users: Identifiable, Codable, FuzzySearchable {
+    var id: Int
+    var name: String
+    var description: String
+ 
+    var searchableString: String {
+        return name
+    }
+}
+```
+
+And now we can just use the fuzzy search function:
+```swift
+@State private var searchText: String = ""
+var users: [Users] = getUsers()
+var sortedUsers: [Users] {
+    if searchText.isEmpty {
+    	return users
+    } else {
+    	return users.fuzzySearch(query: searchText) map {
+            $0.item
+        }
+    }
+}
+```
+
+## Conclusion
```

Comment:
The spacing for this paragraph looks odd when rendered on the blog. Maybe collate some sentences that are broken into multiple lines right now.

Reply:
Should look better now